### PR TITLE
Allow Operator Generated bootstrap token

### PR DESCRIPTION
**Pull request by lhaig**
This PR provides an operator the ability to provide a generated management token which is then used to bootstrap the cluster ACLs.

TODO:
- [x] Tests
- [x] - Input Token Validation
- [x] - Remove hacky code [here](https://github.com/hashicorp/nomad/blob/16ec9038f56487af7ca136cc25845717da71aef2/command/acl_bootstrap.go#L91)

### CODE REVIEW

File: command/agent/acl_endpoint.go [hunk 1]
```diff
@@ -316,3 +319,25 @@ func (s *HTTPServer) ExchangeOneTimeToken(resp http.ResponseWriter, req *http.Re
 	setIndex(resp, out.Index)
 	return out, nil
 }
+
+// parseACLSecrets extracts a map of k/v pairs from the ACL secrets
+// header. Silently ignores invalid secrets
+func parseACLSecrets(req *http.Request) structs.ACLSecrets {
```
Comment (tgross) [inline command/agent/acl_endpoint.go:325]: See my comment in the `api` package about this API.

**Review by tgross — CHANGES_REQUESTED** [review]
Hi @lhaig! This is a great first pass. Most of the implementation looks pretty good, but I've left some comments about the API.

There's currently a failing test you'll want to fix when you go thru and add tests, and make sure you add documentation to the [bootstrap command docs](https://github.com/hashicorp/nomad/blob/main/website/content/docs/commands/acl/bootstrap.mdx)

File: nomad/acl_endpoint.go [hunk 2]
```diff
@@ -396,6 +397,12 @@ func (a *ACL) Bootstrap(args *structs.ACLTokenBootstrapRequest, reply *structs.A
 		Global:     true,
 		CreateTime: time.Now().UTC(),
 	}
+
+	//if a token has been passed in from the comand line overwrite the created one.
+	if providedTokenID != "" {
+		args.Token.SecretID = providedTokenID
+	}
```
[the remark this answers is not in the document]
Comment (tgross) [reply nomad/acl_endpoint.go:404]: It's a bit of a judgement call. We're only going to need it in this one place, so if the logic is simple enough that it doesn't need to be independently unit-tested, I'd say just have it here in-line. But if you find that it's getting complicated, we could have it as a method on `structs.ACLToken`.

[the remark this answers is not in the document]
Comment (lhaig) [reply nomad/acl_endpoint.go]: There is a validator function already and I have used that.

File: api/acl.go [hunk 3]
```diff
@@ -72,9 +73,39 @@ func (c *Client) ACLTokens() *ACLTokens {
...
 
+// CSISecrets contain optional additional credentials that may be needed by
+// the Bootstrap Command.
+// TODO
+// These values will be redacted when reported in the
+// API or in Nomad's logs
+type ACLSecrets map[string]string
+type BootstrapRequest struct {
+	Secrets ACLSecrets
+	WriteRequest
```
[the remark this answers is not in the document]
Comment (lhaig) [reply api/acl.go:85]: I reworked the code as you explained it was not needed for the request.

**Comment by lhaig** [comment]
Hi @tgross,
I was wondering if I need to do anything to the API endpoint to prepare this for use with the terraform provider?
I am not sure what would be the best way to add this as an option

File: api/acl_test.go [hunk 4]
```diff
@@ -269,3 +269,35 @@ func TestACL_OneTimeToken(t *testing.T) {
...
+		c.ACL.Enabled = true
+	})
+	defer s.Stop()
+	at := c.ACLTokens()
+
+	bootkn := &BootstrapRequest{}
+	bootkn.Secret = "2b778dd9-f5f1-6f29-b4b4-9a5fa948757a"
+	// Bootstrap with Valid token
+	out, wm, err := at.Bootstrap(bootkn, nil)
+	assert.Nil(t, err)
```
Comment (tgross) [inline api/acl_test.go:300]: This gives a bit nicer error message on test failure:

```suggestion
	assert.NoError(t, err)
```

File: api/acl.go [hunk 5]
```diff
@@ -72,9 +72,31 @@ func (c *Client) ACLTokens() *ACLTokens {
 	return &ACLTokens{client: c}
 }
 
-// Bootstrap is used to get the initial bootstrap token
-func (a *ACLTokens) Bootstrap(q *WriteOptions) (*ACLToken, *WriteMeta, error) {
```
Comment (tgross) [inline api/acl.go:76]: I missed this the first time through, but changing the signature here breaks backwards compatibility for the Nomad Go API (used by Terraform and third-parties) even though the HTTP API change has been done in a backwards compatible way. So rather than changing the `Bootstrap` method signature, we'll need to:

* Add a new method named `BootstrapOpts` (I'm 😐  about the name but that aligns with other examples)
* Add a comment to the old `Bootstrap` method indicating that the old method is deprecated and will be removed in Nomad 1.5.0.
* Update the code you've changed in this PR to use the new method.

File: command/acl_bootstrap.go [hunk 6]
```diff
@@ -82,8 +88,12 @@ func (c *ACLBootstrapCommand) Run(args []string) int {
 		return 1
 	}
 
+	// Take the provided ACL token and prepare it for the query.
+	btkn := api.BootstrapRequest{}
+	btkn.Secret = bootstraptoken
```
Comment (tgross) [inline command/acl_bootstrap.go:93]: Style and consistency nit: this is a small struct so it's very readable to instantiate and populate its fields all at once:

```suggestion
	btkn := &api.BootstrapRequest{Secret: bootstraptoken}
```

File: nomad/acl_endpoint.go [hunk 7]
```diff
@@ -396,6 +398,16 @@ func (a *ACL) Bootstrap(args *structs.ACLTokenBootstrapRequest, reply *structs.A
 		Global:     true,
 		CreateTime: time.Now().UTC(),
 	}
+
+	// if a token has been passed in from the API overwrite the generated one.
+	if providedTokenID != "" {
+		if helper.IsUUID(providedTokenID) {
+			args.Token.SecretID = providedTokenID
+		} else {
+			return structs.NewErrRPCCodedf(400, "invalid acl token: %v", err)
```
Comment (tgross) [inline nomad/acl_endpoint.go:407]: `err` is always `nil` at this point because our conditional check for UUID doesn't return an error:

```suggestion
			return structs.NewErrRPCCodedf(400, "invalid acl token")
```

File: api/acl.go [hunk 8]
```diff
@@ -72,9 +72,31 @@ func (c *Client) ACLTokens() *ACLTokens {
...
+// TODO
+// These values will be redacted when reported in the
+// API or in Nomad's logs
```
Comment (tgross) [inline api/acl.go:77]: We don't log request bodies or headers, even in the audit log. The token _isn't_ redacted in the API for responses where the user needs to have the token already. Ex. [Read Self Token](https://www.nomadproject.io/api-docs/acl-tokens#read-self-token). So nothing to do here.

```suggestion
```

File: command/acl_bootstrap_test.go [hunk 9]
```diff
@@ -76,3 +78,67 @@ func TestACLBootstrapCommand_NonACLServer(t *testing.T) {
...
+	ui := cli.NewMockUi()
+	cmd := &ACLBootstrapCommand{Meta: Meta{Ui: ui, flagAddress: url}}
+
+	code := cmd.Run([]string{"-address=" + url, "-bootstrap-token=" + mockToken.SecretID})
+	assert.Equal(0, code)
+
+	out := ui.OutputWriter.String()
+	if !strings.Contains(out, mockToken.SecretID) {
+		t.Fatalf("expected "+mockToken.SecretID+" output, got: %s", out)
+	}
```
Comment (tgross) [inline command/acl_bootstrap_test.go:111]: We lean pretty heavily on testify's great assertions:

```suggestion
	assert.Contains(t, out, mockToken.SecretID) {
```

(Applies to the test below as well.)

File: command/acl_bootstrap_test.go [hunk 10]
```diff
@@ -76,3 +78,67 @@ func TestACLBootstrapCommand_NonACLServer(t *testing.T) {
...
+		c.ACL.PolicyTTL = 0
+	}
+
+	// create a valid token
+	mockToken := mock.ACLToken()
+
+	srv, _, url := testServer(t, true, config)
+	defer srv.Shutdown()
+
+	assert.Nil(srv.RootToken)
```
Comment (tgross) [inline command/acl_bootstrap_test.go:100]: If this assertion fails (which it never should), then the whole rest of the test is invalid. This is a good use case for using `require` instead of `assert`:

```suggestion
	require.Nil(t, srv.RootToken)
```

(Applies to the test below as well.)

File: command/acl_bootstrap_test.go [hunk 11]
```diff
@@ -76,3 +78,67 @@ func TestACLBootstrapCommand_NonACLServer(t *testing.T) {
 	out := ui.OutputWriter.String()
 	assert.NotContains(out, "Secret ID")
 }
+
+// Attempting to bootstrap the server with an operator provided token should
+// return the same token in the result.
+func TestACLBootstrapCommand_WithOperatorBootstrapToken(t *testing.T) {
+	ci.Parallel(t)
+	assert := assert.New(t)
```
Comment (tgross) [inline command/acl_bootstrap_test.go:86]: We're trying to remove this aliasing from the code base (same for the tests below). I thought it was picked up by a lint though, so that may need to be tweaked.

```suggestion
```

File: website/content/docs/commands/acl/bootstrap.mdx [hunk 12]
```diff
@@ -25,10 +28,11 @@ The `acl bootstrap` command requires no arguments.
 
 - `-json` : Output the bootstrap response in JSON format.
 - `-t` : Format and display the deployments using a Go template.
+- `-bootstrap-token` : Provide an operator generated management token.
```
Comment (tgross) [inline website/content/docs/commands/acl/bootstrap.mdx:31]: In addition to this document, we should add a line about the bootstrap token header to the [HTTP API docs for ACL tokens](https://github.com/hashicorp/nomad/blob/main/website/content/api-docs/acl-tokens.mdx)

**Review by tgross — CHANGES_REQUESTED** [review]
@lhaig this is looking great! Just a few more things to clean up.

Also, we'll need a changelog entry file. You can add a file at `.changelog/12520.txt` and either use the text of another `improvement` entry as an example, or use the handy [new-cl]( https://github.com/shoenig/new-cl) tool.


> I was wondering if I need to do anything to the API endpoint to prepare this for use with the terraform provider?
> I am not sure what would be the best way to add this as an option.

The Terraform provider literally just lifts this `api` package and uses it directly, so there's nothing special we have to do to add support in the `api` package for it. Fortunately / unfortunately the Terraform provider doesn't currently have a resource for bootstrapping the ACL tokens, so we need to make a new resource that could do so.

File: api/acl.go (hunk 5 above)
Comment (lhaig) [reply api/acl.go]: I have implemented this in the latest push

File: website/content/api-docs/acl-tokens.mdx [hunk 13]
```diff
@@ -11,7 +11,7 @@ For more details about ACLs, please see the [ACL Guide](https://learn.hashicorp.
 
 ## Bootstrap Token
 
-This endpoint is used to bootstrap the ACL system and provide the initial management token.
+This endpoint is used to bootstrap the ACL system and provide the initial management token. An operator created token can be provided in a header called `X-Nomad-Bootstrap-Token` to bootstrap the cluster if required. If no header is provided the cluster will return a generated management token.
```
[the remark this answers is not in the document]
Comment (lhaig) [reply website/content/api-docs/acl-tokens.mdx:14]: Good call, They should know this but it is better to call this out

File: command/acl_bootstrap_test.go [hunk 14]
```diff
@@ -76,3 +78,61 @@ func TestACLBootstrapCommand_NonACLServer(t *testing.T) {
 	out := ui.OutputWriter.String()
 	assert.NotContains(out, "Secret ID")
 }
+
+// Attempting to bootstrap the server with an operator provided token should
+// return the same token in the result.
+func TestACLBootstrapCommand_WithOperatorBootstrapToken(t *testing.T) {
+	ci.Parallel(t)
+	// assert := assert.New(t)
```
[the remark this answers is not in the document]
Comment (lhaig) [reply command/acl_bootstrap_test.go:86]: fixed :-)

File: nomad/structs/structs.go [hunk 15]
```diff
@@ -11914,6 +11914,7 @@ type ACLTokenDeleteRequest struct {
 type ACLTokenBootstrapRequest struct {
 	Token      *ACLToken // Not client specifiable
 	ResetIndex uint64    // Reset index is used to clear the bootstrap token
+	Secret     string
```
Comment (schmichael) [inline nomad/structs/structs.go:11917]: Let's call this `BootstrapSecret`, it's redundant with the struct name, but the fact that we have a weird `Token` field in here makes it non-obvious that this field *is* client-settable.

File: command/agent/acl_endpoint.go [hunk 16]
```diff
@@ -138,8 +138,11 @@ func (s *HTTPServer) ACLTokenBootstrap(resp http.ResponseWriter, req *http.Reque
 		return nil, CodedError(405, ErrInvalidMethod)
 	}
 
+	secret := req.Header.Get("X-Nomad-Bootstrap-Token")
```
Comment (schmichael) [inline command/agent/acl_endpoint.go:141]: Why are we using a header instead of the existing `Secret` (or `BootstrapSecret` as I suggest elsewhere) field? I don't think there's anywhere else we *set* data using a header. :thinking:

File: command/acl_bootstrap.go [hunk 17]
```diff
@@ -30,15 +30,19 @@ Bootstrap Options:
...
 }
 
 func (c *ACLBootstrapCommand) AutocompleteFlags() complete.Flags {
 	return mergeAutocompleteFlags(c.Meta.AutocompleteFlags(FlagSetClient),
 		complete.Flags{
-			"-json": complete.PredictNothing,
-			"-t":    complete.PredictAnything,
+			"-json":            complete.PredictNothing,
+			"-t":               complete.PredictAnything,
+			"-bootstrap-token": complete.PredictAnything,
```
Comment (schmichael) [inline command/acl_bootstrap.go:45]: I feel like putting secrets on the command line is rarely the best option. Maybe @picatz can suggest a good standard approach for securely specifying secrets on a command line.

My thoughts are:
1. Keep this code. I don't think it's such a terrible idea we need to drop it.
2. Accept an `@filename` and/or `-` to read from stdin. Filename seems risky, but stdin might work really well for scripting.
3. `NOMAD_BOOTSTRAP_TOKEN="..."` env var to trigger the same behavior.

Also @picatz if I'm being overly cautious please say so! This feature is a *fantastic* quality of life improvement, and I really don't want to hold it up!

File: command/agent/acl_endpoint.go (hunk 16 above)
Comment (lhaig) [reply command/agent/acl_endpoint.go:141]: I was using the CSI driver as an example when I did the initial experiment.
The reason for this was an easy way to get the token passed from the client to the API that least impacted any other code.

File: nomad/structs/structs.go (hunk 15 above)
Comment (lhaig) [reply nomad/structs/structs.go:11917]: OK will do

File: command/acl_bootstrap.go (hunk 17 above)
Comment (lhaig) [reply command/acl_bootstrap.go:45]: I am happy to add extra options if needed.  I would need to find some examples of how best to do this and it could be implemented.
The standard method of a generated secret is received back in the terminal and so could have been captured in the terminal log as it is.  I would say that the risk is small if someone is running the client to set this. The operator running this command would need to secure the result for an autogenerated token. Our suggestion us always to use this token to create other tokens and then revoke it.
The ideal would be if they used terraform (after I update the provider)

File: command/agent/acl_endpoint.go (hunk 16 above)
Comment (lhaig) [reply command/agent/acl_endpoint.go:141]: @schmichael do you want me to change anything here?

File: command/acl_bootstrap.go (hunk 17 above)
Comment (lhaig) [reply command/acl_bootstrap.go:45]: @picatz @schmichael are there any changes you want me to do here?

Comment (schmichael) [reply command/acl_bootstrap.go:45]: Let's at least add `-bootstrap-token -` (note the trailing `-`) as a signal to read a token from stdin until newline.

That will allow both humans and scripts running the command to at least keep the token out of `ps aux` output.

File: command/agent/acl_endpoint.go (hunk 16 above)
Comment (schmichael) [reply command/agent/acl_endpoint.go:141]: Yes, let's switch to putting the bootstrap token into the request body to follow what most of our other APIs do. Since `ACLTokenBootstrapRequest.BootstrapSecret` already exists you can use it in the CLI and API.

Force-rescheduling a job is a similar example to follow of adding a request body to a command that didn't have one before:
CLI: https://github.com/hashicorp/nomad/blob/v1.3.0/command/job_eval.go#L116-L119
API: https://github.com/hashicorp/nomad/blob/v1.3.0/command/agent/job_endpoint.go#L110-L126

Comment (lhaig) [reply command/agent/acl_endpoint.go:141]: Thanks I will check it out.

File: command/acl_bootstrap.go (hunk 17 above)
Comment (lhaig) [reply command/acl_bootstrap.go:45]: Thank you .

Comment (tgross) [reply command/acl_bootstrap.go:45]: > On that note, I imagine we should enforce some sort of lower-limit for this secret? For example: it needs to be at least 36 characters long.

We're validating that it's a UUID. (ref https://github.com/hashicorp/nomad/pull/12520/files#diff-28b31419c75cf45c1b9bf98fdc46721bb81a71c52e3004e0a01166f4230e561fR402-R409)

Comment (schmichael) [reply command/acl_bootstrap.go:45]: > On that note, I imagine we should enforce some sort of lower-limit for this secret? For example: it needs to be at least 36 characters long.

We enforce the secret is a UUID, which while a bit strange, at least makes `@` detection easy (as its not a valid UUID character) and avoids other issues like complexity checks.

File: command/agent/acl_endpoint.go (hunk 16 above)
Comment (lhaig) [reply command/agent/acl_endpoint.go]: Done

File: .vscode/launch.json [hunk 18]
```diff
@@ -0,0 +1,7 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": []
+}
```
Comment (schmichael) [inline .vscode/launch.json:7]: Mind removing this file and/or adding it to `.gitignore`?

File: api/acl.go [hunk 19]
```diff
@@ -244,3 +262,8 @@ type OneTimeTokenExchangeRequest struct {
 type OneTimeTokenExchangeResponse struct {
 	Token *ACLToken
 }
+
+// Bootstraorequest is used to when operstors priovide an ACL Bootstrap Token
```
Comment (schmichael) [inline api/acl.go:266]:
```suggestion
// BootstrapRequest is used to when operators provide an ACL Bootstrap Token
```

File: command/agent/acl_endpoint.go [hunk 20]
```diff
@@ -138,8 +138,18 @@ func (s *HTTPServer) ACLTokenBootstrap(resp http.ResponseWriter, req *http.Reque
...
 
-	// Format the request
-	args := structs.ACLTokenBootstrapRequest{}
+	var args structs.ACLTokenBootstrapRequest
+
+	if req.ContentLength == 0 {
+		args = structs.ACLTokenBootstrapRequest{
+			BootstrapSecret: "",
+		}
+	} else {
```
Comment (schmichael) [inline command/agent/acl_endpoint.go:147]: You can rely on Go initializing structs to their zero/empty values and simplify this code a bit. No big deal.

```suggestion
	args := structs.ACLTokenBootstrapRequest{}

	if req.ContentLength != 0 {
```

File: website/content/api-docs/acl-tokens.mdx [hunk 21]
```diff
@@ -11,7 +11,10 @@ For more details about ACLs, please see the [ACL Guide](https://learn.hashicorp.
 
 ## Bootstrap Token
 
-This endpoint is used to bootstrap the ACL system and provide the initial management token.
+This endpoint is used to bootstrap the ACL system and provide the initial management token. 
+An operator created token can be provided in a header called `X-Nomad-Bootstrap-Token` 
+to bootstrap the cluster if required. If no header is provided the cluster will return a 
+generated management token. The provided token should be presented in a UUID format.
```
Comment (schmichael) [inline website/content/api-docs/acl-tokens.mdx:17]: This will need to be updated to mention the request body instead of the header.

**Review by schmichael — COMMENTED** [review]
Looking great! Just need the file/stdin reading! Thanks for sticking with this Lance.

File: .gitignore [hunk 22]
```diff
@@ -125,3 +125,6 @@ e2e/remotetasks/input/ecs.vars
 
 # local terraform overrides
 *.auto.tfvars
+
+# vscode config
+.vscode/launch.json
```
Comment (apollo13) [inline .gitignore:130]: Mhm people can add that to their ~/.gitignore and have it apply locally instead of adding arbritrary ignores for every ide/editor out there?

File: command/agent/acl_endpoint.go (hunk 20 above)
Comment (lhaig) [reply command/agent/acl_endpoint.go:147]: Simplify Code

File: .gitignore (hunk 22 above)
Comment (tgross) [reply .gitignore:130]: We have other commonly used editor files in the .gitignore. It reduces the amount of churn, particularly from external contributors. I'd leave it in. (But maybe move it up so that it's group with `./idea`, `.iml`, etc.)

File: command/acl_bootstrap.go (hunk 17 above)
Comment (lhaig) [reply command/acl_bootstrap.go:45]: Would it not be better to just remove the -bootstrap-token flag completely and then just accept 
- No args (Standard generated token)
- stdin read from there
- filename to read a file

Comment (schmichael) [reply command/acl_bootstrap.go:45]: Yeah! That works but use `-` as the "filename" for stdin so the cases are:

```
nomad acl bootstrap   # same as today
nomad acl bootstrap - # read from stdin
nomad acl bootstrap root.token # read from file root.token
```

No big deal if you want to keep the flag though. Since folks should only be running this once-per-cluster I don't want to spend too much time optimizing the ergonomics. My main concern is security which we've already covered. :+1:

Comment (lhaig) [reply command/acl_bootstrap.go:45]: It makes it easier to code if I remove the flag. for me at least.

I just need to fix the tests and then I can update the documents and we are good for this iteration.

File: command/acl_bootstrap.go [hunk 23]
```diff
@@ -59,28 +61,48 @@ func (c *ACLBootstrapCommand) Name() string { return "acl bootstrap" }
...
+		case "-":
+			terminalToken, err = ioutil.ReadAll(os.Stdin)
+		default:
+			terminalToken, err = ioutil.ReadFile(file)
+		}
+		if err != nil {
+			c.Ui.Error(fmt.Sprintf("Error reading provided token: %v", err))
+			return 1
+		}
+	}
```
Comment (schmichael) [inline command/acl_bootstrap.go:101]: Wow! Well that approach turned out great!

File: api/acl.go [hunk 24]
```diff
@@ -72,6 +72,7 @@ func (c *Client) ACLTokens() *ACLTokens {
 	return &ACLTokens{client: c}
 }
 
+// This Method is deprecated and will be removed in 1.5
```
[the remark this answers is not in the document]
Comment (lhaig) [reply api/acl.go:75]: @tgross suggested this would be a way forward. 
Having more than one way to do similar thing might be confusing.

File: command/agent/acl_endpoint.go [hunk 25]
```diff
@@ -138,8 +138,14 @@ func (s *HTTPServer) ACLTokenBootstrap(resp http.ResponseWriter, req *http.Reque
 		return nil, CodedError(405, ErrInvalidMethod)
 	}
 
-	// Format the request
-	args := structs.ACLTokenBootstrapRequest{}
+	var args structs.ACLTokenBootstrapRequest
+
+	if req.Body != nil {
```
[the remark this answers is not in the document]
Comment (lhaig) [reply command/agent/acl_endpoint.go:143]: The reason i went for body is that the content length is not reliable and sometimes returns 0 even when the body contains the full token.
So I went for what I thought would be a reliable option.

File: website/content/api-docs/acl-tokens.mdx [hunk 26]
```diff
@@ -51,6 +54,39 @@ $ curl \
...
+  "BootstrapSecret": "2b778dd9-f5f1-6f29-b4b4-9a5fa948757a"
+}
+```
+
+### Sample Request With Operator Token
+
+```shell-session
+$ curl \
+    --request POST \
+    --data @root.token \
```
[the remark this answers is not in the document]
Comment (lhaig) [reply website/content/api-docs/acl-tokens.mdx]: I could have sworn that I changed that.

File: command/agent/acl_endpoint.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (lhaig) [reply command/agent/acl_endpoint.go:143]: @schmichael do you want me to change anything here?

[the remark this answers is not in the document]
Comment (schmichael) [reply command/agent/acl_endpoint.go:143]: Yes, `Reqest.Body` [cannot be nil in this server context](https://pkg.go.dev/net/http#Request), so this check is never hit.

`nomad acl bootstrap` now _always_ sends a body, which is fine, but means if you're testing with the CLI it will always work. Your `command/agent/acl_endpoint_test.go` test doesn't test the empty body case.

If you add a test case for an empty body you will see that prior to your change an empty body to this HTTP API works, but after your change the HTTP API returns a 400 on an empty body.

You can also test this with curl:

```
# Should work but fails:
~$ curl -v -XPOST localhost:4646/v1/acl/bootstrap
> POST /v1/acl/bootstrap HTTP/1.1
> Host: localhost:4646
> User-Agent: curl/7.68.0
> Accept: */*
>
< HTTP/1.1 400 Bad Request
< Vary: Accept-Encoding
< Date: Tue, 31 May 2022 21:45:53 GMT
< Content-Length: 21
< Content-Type: text/plain; charset=utf-8
<
Request body is empty
```

File: command/agent/acl_endpoint_test.go [hunk 27]
```diff
@@ -221,6 +221,49 @@ func TestHTTP_ACLTokenBootstrap(t *testing.T) {
...
+
+		// Make the HTTP request
+		req, err := http.NewRequest("PUT", "/v1/acl/bootstrap", buf)
+		if err != nil {
+			t.Fatalf("err: %v", err)
+		}
+
+		// Since we're not actually writing this HTTP request, we have
+		// to manually set ContentLength
+		req.ContentLength = -1
```
Comment (schmichael) [inline command/agent/acl_endpoint_test.go:246]: Wow this took me a while to figure out: since we're never actually writing this HTTP request, the stdlib code that sets ContentLength is never set!

Therefore in tests we need to set ContentLength to some nonzero value since tests do not use a real HTTP server.

**Review by schmichael — APPROVED** [review]
Looks good to me! Sorry our test harness breaks the ContentLength check! What a hassle.

~~I'm not sure why the lint tests are failing. They pass for me locally. Maybe someone else on the team will have an idea.~~

_Update: Fixed lint tests! I think I just needed to upgrade hclfmt._

File: command/agent/acl_endpoint_test.go (hunk 27 above)
Comment (lhaig) [reply command/agent/acl_endpoint_test.go:246]: I am glad it is finally solved it kept me up a few nights :-)

**Review by tgross — APPROVED** [review]
LGTM! Kudos for sticking this one out @lhaig! Reviews for a big feature can sometimes be a good bit of work but Nomad is going to be that much better for this PR. Thanks!
